### PR TITLE
`UnitControl`: Revamp support for changing unit by typing

### PR DESCRIPTION
**Pull request by stokesman**
## What?
Revamps the type-to-change-units feature in `UnitControl` and removes it's dependency on the `isPressEnterToChange` prop being `true`. ~~Also removes `BoxControl`’s defaulting of `isPressEnterToChange` to `true` in order for it to send changes instantaneously (as opposed to now how it waits for either <kbd>ENTER</kbd> key press or a `blur` event)~~.

## Why?
~~Betters UX consistency in regard to inputs that produce changes instantly. Unlike all other text inputs in Gutenberg, `BoxControl`-based ones do not produce changes upon input. It seems the only reason for this difference is to enable the type-to-change-units feature which requires it.~~ With this PR, the feature works even when the mode of these controls set to instantaneous (`isPressEnterToChange={ false }`)

The feature itself may also be found to work better but testing will tell. It's new implementation is certainly less complex and better from a maintenance perspective.

Related:
* Issue with consistency noted https://github.com/WordPress/gutenberg/issues/30222#issue-840656311
* We've also recently merged #39109 that seems to have been motivated by the urge to alleviate this.

## How?
* Removal of the logic that currently supports the feature.
* Addition of simpler logic that merely moves focus to the unit select when a key that matches the first character of a unit. 
* ~~Change of the default `isPressEnterToChange` of `BoxControl` to `false`.~~

## Testing Instructions
1. Play with a `UnitControl` in Storybook or Gutenberg.
2. Verify that the typing a unit from the text input changes the units.

## Screenshots or screencast
https://user-images.githubusercontent.com/9000376/157374684-111cead1-3c3f-4021-9a60-7285d4a3c878.mp4

### CODE REVIEW

File: packages/components/src/unit-control/index.tsx
Comment (stokesman) [inline packages/components/src/unit-control/index.tsx]: As far as I can tell the use of `'text'` is only to allow input of any character in Chrome (other common browsers don't limit input). With this PR that's no longer a concern. Removing this will default `type` to `'number'` and have the possible benefit of easier validation since text inputs don't factor `min`, `max`, or `step` into validity.

**Comment by stokesman** [comment]
I'm not soliciting code reviews since this first needs vetting from a UX perspective. It would be awesome to have some folks test this out. I think @andrewserong and @aaronrobertshaw may be interested since they've made related issues/PRs; @ciampo and @mirka since this is components territory.

Gutenberg.run is functioning again (as of this comment) so this [PR can be tested there](http://gutenberg.run/39303). To save a couple seconds, here's a Group block to that can be pasted to test with its padding `BoxControl`:
```HTML
<!-- wp:group {"backgroundColor":"secondary"} -->
<div class="wp-block-group has-secondary-background-color has-background"></div>
<!-- /wp:group -->
```

Probably the first potential stopper—is it verboten to move focus in this manner?

File: packages/components/src/box-control/test/index.js
Comment (ciampo) [inline packages/components/src/box-control/test/index.js]: Now that we have `@testing-library/user-event` added to the repo, it'd be great if we could rewrite these tests (although definitely a separate task from this PR)

File: packages/components/src/unit-control/index.tsx
Comment (ciampo) [inline packages/components/src/unit-control/index.tsx]: I believe we should pass `stateReducerProp` directly?

File: packages/components/src/unit-control/index.tsx
Comment (ciampo) [inline packages/components/src/unit-control/index.tsx]: We will need to rebase this PR and undo this change once #39589 gets merged

File: packages/components/src/unit-control/index.tsx
Comment (ciampo) [reply packages/components/src/unit-control/index.tsx]: > As far as I can tell the use of `'text'` is only to allow input of any character in Chrome

This is an interesting change, to the point that I'd almost split it to a separate PR so that we can focus on testing it in isolation - WDYT @mirka ?

File: packages/components/src/unit-control/index.tsx
Comment (ciampo) [inline packages/components/src/unit-control/index.tsx]: Nit: The `| null` part is not need when passing `null` as the default value

```suggestion
	const refInputSuffix = useRef< HTMLSelectElement >( null );
```

File: packages/components/src/unit-control/index.tsx
Comment (ciampo) [inline packages/components/src/unit-control/index.tsx:167]: > Probably the first potential stopper—is it verboten to move focus in this manner?

I don't know the answer myself at the moment, but we definitely need to assess if moving focus this way works well for folks relying on assistive technology

**Review by ciampo — COMMENTED** [review]
Hey @stokesman , thank you for working on this exploration!

If we agreed on moving forward with the changes in this PR, what usecases are left for keeping the `isPressEnterToChange` prop on `UnitControl` / `NumerControl` / `InputControl`?

File: packages/components/src/unit-control/index.tsx
Comment (stokesman) [reply packages/components/src/unit-control/index.tsx]: Ha! I think I removed it from the props too but then missed it on a rebase. Now I think you're right. It wouldn't hurt to leave that door open.

File: packages/components/src/unit-control/index.tsx
Comment (mirka) [reply packages/components/src/unit-control/index.tsx]: Yeah, let's test it separately 👍

File: packages/components/src/unit-control/index.tsx
Comment (mirka) [reply packages/components/src/unit-control/index.tsx:167]: Wow, I played around with it and I'm tempted to say it's acceptable, despite the focus moving 🤔 I think a big reason for that is because it's only moving to the next tabbable, and also the "type in the unit in the number field" is a hidden feature that only advanced users know in the first place. (Do we know how often people use this feature? I don't think I would ever discover it.)

But ideally, if we could accomplish this without a focus jump that might be even better. I guess we can't do something like remember the successive keydowns and wait for a complete match, because there could be units like `m` and `mm` ☹️

File: packages/components/src/unit-control/index.tsx
Comment (mirka) [inline packages/components/src/unit-control/index.tsx:161]: Should we also handle the `onPaste` case?

**Comment by alexstine** [comment]
@mirka How exactly do I test this? Not sure where BoxControl is used.

File: packages/components/src/unit-control/index.tsx
Comment (aaronrobertshaw) [reply packages/components/src/unit-control/index.tsx:161]: Handling `onPaste` would be nice. I've seen a few mentions lately of values being copied and pasted whether that is between controls or from a reference stylesheet.

**Comment by stokesman** [comment]
Hi @alexstine, BoxControl is likely most often encountered as the Padding control in the Dimensions panel in the block settings for a few blocks like Group and Columns. Unless you've already checked this branch out the easiest way to test would be at [gutenberg.run/39303](http://gutenberg.run/39303). If it helps, here's a group block to paste in to save a couple seconds:
```html
<!-- wp:group {"backgroundColor":"secondary"} -->
<div class="wp-block-group has-secondary-background-color has-background"></div>
<!-- /wp:group -->
```

Also I believe the most crucial aspect of this to get your feedback on is the feature which allows changing the units of the control by typing the first character of one of the units while the quantity input is still focused. In this PR it will move focus to the select element immediately. That's opposed to way it works in trunk which doesn't move focus but requires that you type the units in and press enter or blur the input.

**Comment by alexstine** [comment]
@stokesman Here is what I did.
1. Insert a group block.
2. Tab to the field labeled Padding.
3. Enter the number 5.
4. The number was entered with no focus change.

I tested the gutenberg.run link.

Is something else supposed to happen?

Thanks.

File: packages/components/src/unit-control/index.tsx
Comment (aaronrobertshaw) [inline packages/components/src/unit-control/index.tsx]: This might be a bit of a contrived edge case but should we make this regex case insensitive? 

Without that, I can get an `E` to be entered into the input using either capslock or holding shift. It also means if capslock is on for some reason the unit change doesn't work.

Screen Shot 2022-03-23 at 11 22 41 am

```suggestion
		return [ list, new RegExp( `^(?:${ firstCharacters })$`, 'i' ) ];
```

File: packages/components/src/unit-control/index.tsx
Comment (aaronrobertshaw) [reply packages/components/src/unit-control/index.tsx:167]: I also found this to be quite an intuitive experience despite the focus moving. I'm happy though to defer to others who have greater experience with these components and accessibility requirements.

The point regarding this hidden feature not being that well known is a good one.

**Comment by stokesman** [comment]
Thanks for testing @alexstine and for your steps. I've changed step 4.
 1.  Insert a group block.
 2. Tab to the field labeled Padding.
 3. Enter the number 5.
 4. Enter one of the following characters: % e r v

By following that you should experience the focus change we're trying to scrutinize.

**Comment by alexstine** [comment]
@stokesman I am happy with this. If a user understands enough to enter one of those characters, I think it only makes sense to move to the select field after.

Thanks.

File: packages/components/src/unit-control/index.tsx
Comment (stokesman) [reply packages/components/src/unit-control/index.tsx:161]: That's a good point. It does seem like a bit of an edge case given that you can't copy a string with the units from the control itself but I'd not be opposed to giving it a try.

File: packages/components/src/unit-control/index.tsx
Comment (stokesman) [reply packages/components/src/unit-control/index.tsx]: This is a great point. I think case-insensitive makes sense.

One thing I also wanted to consider was moving the focus when any character besides a numeral is input. But the most obvious win for that may just be simpler code and I'm not sure that it helps UX.

File: packages/components/src/unit-control/index.tsx
Comment (ciampo) [reply packages/components/src/unit-control/index.tsx:167]: I agree with the points made above. This behaviour seems intuitive to me, and only changes what was previously a "power user" niche feature. It also seems like [we don't have any major concern](https://github.com/WordPress/gutenberg/pull/39303#issuecomment-1075822433) from the accessibility point of view.

> But ideally, if we could accomplish this without a focus jump that might be even better. I guess we can't do something like remember the successive keydowns and wait for a complete match, because there could be units like m and mm

I'm not sure how feasible this is without introducing a lot of complexity to a component that is quite complex already

**Comment by ciampo** [comment]
I asked this before, but it likely got ignored because of every other comment that I left: If we agreed on moving forward with the changes in this PR, what use-cases are left for keeping the `isPressEnterToChange` prop on `UnitControl` / `NumerControl` / `InputControl`?

File: packages/components/src/unit-control/index.tsx
Comment (mirka) [reply packages/components/src/unit-control/index.tsx:167]: > I'm not sure how feasible this is without introducing a lot of complexity to a component that is quite complex already

Agreed. Given that everyone here thinks the behavior is intuitive, let's move forward with the current implementation 👍

**Comment by stokesman** [comment]
I hadn't forgotten about that question @ciampo. I couldn't think of a use case and thought maybe with a little time an idea would come to me. Yesterday, I searched in this branch did not see any. A search on plugins with [WP Directory](https://wpdirectory.net/search/01FYSAS5AMGN799FRVVYQ8DB9D) turns up some apparent use but I don't get a clear idea of the use cases by looking at those results. I'm thinking I might test a few of those plugins to try and find out because I'm curious.

Also thank you for the quick review!

File: packages/components/src/box-control/test/index.js
Comment (stokesman) [reply packages/components/src/box-control/test/index.js]: I did forget to address this one. Agreed!

**Comment by ciampo** [comment]
> I couldn't think of a use case and thought maybe with a little time an idea would come to me.

I also can't personally think about any other reason. I found [this comment](https://github.com/WordPress/gutenberg/pull/25933#issuecomment-705550921) which confirms our theory.

> A search on plugins with [WP Directory](https://wpdirectory.net/search/01FYSAS5AMGN799FRVVYQ8DB9D) turns up some apparent use but I don't get a clear idea of the use cases by looking at those results. I'm thinking I might test a few of those plugins to try and find out because I'm curious.

That would definitely be a good idea, let us know if/when you have any findings! It'd be nice if we could (of course, separately to this PR) simplify `InputControl` and derived components a little bit by removing this prop.

> Also thank you for the quick review!

Don't mention it! Thank you for your continuous efforts in contributing to the components package!

Let us know once this PR is ready for another round of reviews

File: packages/components/src/unit-control/index.tsx
Comment (stokesman) [reply packages/components/src/unit-control/index.tsx:161]: A wild `onPaste` handling branch appears #40168. I don't think it’s actually wild but I'd guess we might want to add unit tests for it which I've yet to try. I made it as a separate PR because it seemed non-trivial enough to warrant that.

File: packages/components/src/unit-control/index.tsx
Comment (stokesman) [reply packages/components/src/unit-control/index.tsx:167]: >But ideally, if we could accomplish this without a focus jump that might be even better.

For posterity, I can verify that it can work but does add some complexity. It was how I first approached this and I've since revisited it because I realized that it already enables pasting so it might have a comparable level of complexity to this approach with `onPaste` handling added. Even with some success in simplifying it in my latest attempt it comes out looking complex compared to this. I have an inkling that it could be simplified further but would first require some overhaul to `InputControl` to allow for more flexible extensibility.

File: packages/components/src/unit-control/index.tsx
Comment (stokesman) [reply packages/components/src/unit-control/index.tsx]: I've made sure to keep the behavior from #39589 (as confirmed by the tests added there still passing). I didn't add the prop back here as it can pass through as part of `props`. The unit select takes it up like so:
https://github.com/WordPress/gutenberg/blob/9a0a708bfbb14ec78ffa9204c7eaffc3a2d21fc7/packages/components/src/unit-control/index.tsx#L181

If you'd prefer that `onBlur` is destructured from `unitControlProps` it'd be fine by me.

**Comment by stokesman** [comment]
I've gone ahead and marked this as ready and also made #40168 for the `onPaste` handling. If that is considered a must have for this PR then I guess that'll need review/approval before this one.

File: packages/components/src/unit-control/index.tsx
Comment (stokesman) [reply packages/components/src/unit-control/index.tsx]: I've reverted that from this PR but while working on the paste handling (in #40168 ) I found that robust/conventional support will require that the type be `text`. I left a review comment on that PR that may help elucidate.

**Comment by alexstine** [comment]
Should this one be closed?

**Comment by ciampo** [comment]
This PR has diverged a bit from `trunk` — the `BoxControl` component has been refactored to TypeScript (and its tests to `user-event`).

I'm not sure exactly why we lost momentum on this PR, probably it just got overlooked unintentionally.

To better understand the next steps, we should assess if this behaviour is something that we'd like to have — @alexstine , from your [previous comment](https://github.com/WordPress/gutenberg/pull/39303#issuecomment-1075822433), it looks like the new proposed behavior was viable under the accessibility POV.

It may be good to also hear @aaronrobertshaw 's opinion again, and maybe even @jameskoster 's (for Jay: you can find a high-level overview of the proposed changes in the PR description, including a video)

**Comment by alexstine** [comment]
This did not introduce any accessibility concerns. I am happy with it.

**Comment by aaronrobertshaw** [comment]
Thanks for the ping 👍 

I've taken another quick look at this PR and trunk. There does appear to me to still be some value in resurrecting this PR as it would address the [original issue](https://github.com/WordPress/gutenberg/issues/30222#issue-840656311). It's unfortunate that this didn't land before the TypeScript refactors but I'm hopeful it won't be to time-consuming to resolve conflicts or port the changes into a fresh PR.

I'm a little short on bandwidth at the moment but would be happy to help with reviews etc.

**Comment by jameskoster** [comment]
This seems like a nice little enhancement for power users 👍

**Comment by ciampo** [comment]
@stokesman , would you have capacity to resurrect this PR ? If conflicts are too entangled to solve quickly, we can also open a new PR superseding the current one.

**Comment by stokesman** [comment]
> would you have capacity to resurrect this PR ?

I should be able to make some time later this week. If not, next week for sure.

**Comment by stokesman** [comment]
So this branch is back to being free of conflicts. Picking this back up, the first question I've been occupied with was whether this PR should be changing the default behavior of `BoxControl`. It seems right as far as promoting consistency with other components and general UX. On the other hand, I think we're now treating all components as stable so it’s not strictly okay (yet I'd imagine it’s fine because it's hard to imagine it creating code breakage).

Anyway, the `BoxControl` change is low stakes in the context of Gutenberg because `BoxControl` has almost vanished from use. I went ahead and reverted those changes for now since I figured it might make for a bit less to think about in reviewing this. It would be easy to bring them back or do it in a separate PR.

Now it awaits (patiently) a review.

EDIT: I was too hasty reverting all `BoxControl` changes 😬.  The unit tests do require some of the updates that I'd made so I'll get that fixed up soon but I'm hitting the hay for now.

**Review by ciampo — COMMENTED** [review]
Thank you for iterating on this again, @stokesman.

I'll wait until tests are passing to give a good look at this PR again.

Hoping that everything tests well and feels good, it is great to see `UnitControl`'s implementation getting simpler. With this PR merged, the only usage of the `isPressEnterToChange` prop would to be [in the `DateTime` component](https://github.com/WordPress/gutenberg/blob/8dee3e6764da100b20ad9abd41af5df086b5f42f/packages/components/src/date-time/time/index.tsx) — I wonder if we could look at gradually removing this functionality from `InputControl` & co altogether, which would lead to a further great code simplification.

File: packages/components/src/box-control/test/index.tsx
Comment (ciampo) [inline packages/components/src/box-control/test/index.tsx]: Should we actually test that the focus was moved correctly?

File: packages/components/src/box-control/test/index.tsx
Comment (ciampo) [inline packages/components/src/box-control/test/index.tsx]: Exactly what part of the implementation doesn't work in the test environment? It would be great if we had unit tests for this new behavior

**Review by ciampo — COMMENTED** [review]
Thank you for iterating on this, @stokesman ! (just a quick reminder about adding a CHANGELOG entry)

From a quick glance in Storybook, the component seems to work well, although we definitely need to test this PR well, given how `UnitControl` is used in many places across the editor.

@aaronrobertshaw  could you take another look at `UnitControl` and `BoxControl`, both in isolation and in the editor?

File: packages/components/src/box-control/test/index.tsx
Comment (stokesman) [reply packages/components/src/box-control/test/index.tsx]: In dd0219c coverage was added for that in tests for `UnitControl`:
https://github.com/WordPress/gutenberg/blob/dd0219c0cccae682d6c67a9bf700e36f804bccc9/packages/components/src/unit-control/test/index.tsx#L590-L601

Covering it again in the tests for `BoxControl` seems likely to be unnecessary and not the focus of this particular test.

File: packages/components/src/box-control/test/index.tsx
Comment (stokesman) [reply packages/components/src/box-control/test/index.tsx]: The actual unit switching is entirely up to the browser. The component merely moves the focus and what seems to be de-facto standard behavior of browsers changes the select according to key presses. We could cover it with e2e tests.

**Comment by stokesman** [comment]
I pushed some changes for some minor optimization c51c1bd, code style revision 8225f905248904fe8834133c3331e274f4876176, expanded test coverage ac16dd0 and a CHANGELOG entry.

> With this PR merged, the only usage of the `isPressEnterToChange` prop would to be [in the DateTime component](https://github.com/WordPress/gutenberg/blob/8dee3e6764da100b20ad9abd41af5df086b5f42f/packages/components/src/date-time/time/index.tsx)…

I'd reverted its removal from `BoxControl` in this PR before that comment was made so I just wanted to clarify that. I'm pretty sure removing from `BoxControl` should be inconsequential but I thought to avoid that consideration from this PR.

File: packages/components/src/box-control/test/index.tsx
Comment (stokesman) [reply packages/components/src/box-control/test/index.tsx]: ac16dd0 updates the test mentioned in the comment above to run for each of the default units of `UnitControl`.

**Review by ciampo — APPROVED** [review]
Rebased and made a bunch of improvements to unit tests, I think this PR is good to go 🎉 

Thank you @stokesman for your work on this one !

There seem to be a few follow-up tasks:

- investigating the removal of all `isPressEnterToChange` usages from existing components (ie. `BoxControl`, `DateTimePicker`) (this should also help us with #30222)
- depending on how the previous task goes, we could then consider deprecating the `isPressEnterToChange` prop, and the behavior associated with it
- add e2e tests to check the expected unit selection behavior, which can't be currently tested with `js-dom` as it seems like it doesn't support changing `select` element values using the keyboard
- enhance the `onPaste` behavior to also support unit change (work started in #40168)

@stokesman , would you have any capacity to work on the above tasks? 🙏

**Comment by stokesman** [comment]
Thanks for moving this one through Marco!

> would you have any capacity to work on the above tasks?

I can probably chip away at those. The first two would hopefully be pretty straightforward.

> add e2e tests to check the expected unit selection behavior…

Because they'd cover behavior that the library doesn't control I wouldn't choose to write them. I mentioned the idea to say it could be done but not that it should be 🤷.

Thanks again to all who tested and reviewed 🙌 ❤️ I'm fond of this little change.

**Pull request merged**
